### PR TITLE
Fixes #33108 - Add start_at for job recurrence

### PR DESCRIPTION
**Pull request by xprazak2**
Needs https://github.com/theforeman/foreman-tasks/pull/643

### CODE REVIEW

File: app/controllers/api/v2/job_invocations_controller.rb [hunk 1]
```diff
@@ -47,6 +47,7 @@ def show
             param :cron_line, String, :required => false, :desc => N_('How often the job should occur, in the cron format')
             param :max_iteration, :number, :required => false, :desc => N_('Repeat a maximum of N times')
             param :end_time, DateTime, :required => false, :desc => N_('Perform no more executions after this time')
+            param :start_at, DateTime, :required => false, :desc => N_('Schedule the first iteration after this time')
```
Comment (adamruzicka) [inline app/controllers/api/v2/job_invocations_controller.rb:50]: This trips up hammer. The value ends up under both `recurrence` and `scheduling` keys in the request being sent to the server.

```
[DEBUG 2021-07-28T14:34:23 API] Params: {
    "job_invocation" => {
                     "inputs" => {
            "playbook" => "asd"
        },
                        "ssh" => {},
                 "recurrence" => {
            "cron_line" => "0 0 * * *",
             "start_at" => "2021-07-31T17:00:00+00:00"
        },
                 "scheduling" => {
            "start_at" => "2021-07-31T17:00:00+00:00"
        },
        "concurrency_control" => {},
               "search_query" => "name = test-00.example.com",
                    "feature" => "ansible_run_playbook",
             "targeting_type" => "static_query"
    }
}
[DEBUG 2021-07-28T14:34:23 API] Headers: {}
[DEBUG 2021-07-28T14:34:23 API] Using authenticator: HammerCLIForeman::Api::InteractiveBasicAuth
[ERROR 2021-07-28T14:34:24 API] 500 Internal Server Error
[DEBUG 2021-07-28T14:34:24 API] {
    "error" => {
        "message" => "ERF42-4606 [Foreman::Exception]: Cannot specify both recurrence and scheduling"
    }
}
```

**Review by adamruzicka — COMMENTED** [review]
Works nicely through API, but hammer can't really deal with it. Maybe keeping recurrence and scheduling parameters kept apart but then looking at both wouldn't be such a bad idea?

**Comment by adamruzicka** [comment]
Also, could you expose this through the ui?

**Comment by xprazak2** [comment]
Hammer should no longer be confused.
Changes for exposing start_at in UI were made in foreman-tasks.

**Comment by adamruzicka** [comment]
Tasks counter part just went in, could you bump dependency on tasks to 5.1.0?

**Comment by xprazak2** [comment]
I bumped the tasks.

**Comment by adamruzicka** [comment]
Test failures seem to be related

**Comment by adamruzicka** [comment]
Needs a rebase

File: app/models/job_invocation_composer.rb [hunk 2]
```diff
@@ -138,18 +138,11 @@ def targeting_params
...
-          :input_type => :cronline,
-          :max_iteration => api_params[:recurrence][:max_iteration],
-          :purpose => api_params[:recurrence][:purpose],
-        }
-      elsif api_params[:scheduling].present?
+      if api_params[:recurrence].present? && api_params[:scheduling].present?
+        recurring_mode_params.merge :start_at_raw => format_datetime(api_params[:scheduling][:start_at])
+      elsif api_params[:recurrence].present? && api_params[:scheduling].nil?
+        recurring_mode_params
+      elsif api_params[:recurrence].nil? && api_params[:scheduling].present?
```
Comment (adamruzicka) [inline app/models/job_invocation_composer.rb:145]: nil checks won't work with hammer
```
[DEBUG 2021-09-07T10:39:02 API] Params: {
    "job_invocation" => {
                     "inputs" => {                                                                     
            "playbook" => "asd"
        },
                        "ssh" => {},
                 "recurrence" => {},               
                 "scheduling" => {                                                                     
            "start_at" => "2021-09-15T17:00:00+00:00"
        },                                        
        "concurrency_control" => {},              
               "search_query" => "name = test-00.example.com",              
                    "feature" => "ansible_run_playbook",                             
                    "ansible" => {},                                                                   
             "targeting_type" => "static_query"                                                        
    }                                      
}
```

**Review by adamruzicka — CHANGES_REQUESTED** [review]
Scheduling to future with recurrence seems to work well
```
# h job-invocation create --feature ansible_run_playbook --inputs playbook="asd" --search-query='name = test-00.example.com' --cron-line '0 0 * * *' --start-at '2021-09-15 17:00:00'
Job invocation 524 created

# h job-invocation info --id 524
---
ID: 524
Description: Ansible - Run playbook with inputs playbook="asd"
Status: queued
Success: N/A
Failed: N/A
Pending: N/A
Missing: 0
Total: N/A
Start: 2021-09-16 00:00:00 UTC
Randomized ordering: false
Job Category: Ansible Playbook
Mode: recurring
Cron line: 0 0 * * *
Recurring logic ID: 58
Hosts:
- Name: test-00.example.com
```

However just scheduling to the future does not
```
# h job-invocation create --feature ansible_run_playbook --inputs playbook="asd" --search-query='name = test-00.example.com' --start-at '2021-09-15 17:00:00'         
Job invocation 525 created

# h job-invocation info --id 525
---
ID: 525
Description: Ansible - Run playbook with inputs playbook="asd"
Status: running
Success: N/A
Failed: N/A
Pending: N/A
Missing: 0
Total: 1
Start: 2021-09-07 08:34:41 UTC
Randomized ordering: false
Job Category: Ansible Playbook
Hosts:
- Name: test-00.example.com
```

Note, the command was run at 10:34 GMT+2 so it just igrnoed start-at completely.

File: app/models/job_invocation_composer.rb (hunk 2 above)
Comment (xprazak2) [reply app/models/job_invocation_composer.rb]: I replaced nil checks with `empty?`, it should take care of the problem with future scheduling as well.

**Review by adamruzicka — APPROVED** [review]
Works well, ACK

**Comment by adamruzicka** [comment]
Thank you @xprazak2 !

**Pull request merged**